internal/cli/base.go

```diff
@@ -397,6 +399,95 @@ func (c *baseCommand) Init(opts ...Option) error {
+	// Check if remote is disabled in the waypoint.hcl
+
+	if !project.RemoteEnabled {
+		log.Debug("Remote operations are disabled for this project - operation cannot occur remotely")
+		return false, nil
+	}
+
+	if project.DataSource == nil {
+		log.Debug("Project has no datasource configured - operation cannot occur remotely")
+		// This is probably going to be fatal somewhere downstream
+		return false, nil
+	}
+
+	var hasRemoteDataSource bool
+	switch project.DataSource.GetSource().(type) {
+	case *pb.Job_DataSource_Git:
+		hasRemoteDataSource = true
+	default:
+		hasRemoteDataSource = false
+	}
+
+	if !hasRemoteDataSource {
+		log.Debug("Project does not have a remote data source - operation cannot occur remotely")
+		return false, nil
+	}
+
+	// We know the project can handle remote ops at this point - but do we have runners?
+
+	runnersResp, err := client.ListRunners(ctx, &pb.ListRunnersRequest{})
+	if err != nil {
+		return false, err
+	}
+	hasRemoteRunner := false
+	for _, runner := range runnersResp.Runners {
+		if !runner.Odr {
+			// NOTE(izaak): There is currently no way to distinguish between a remote runner and a CLI runner.
+			// So if some other waypoint client is performing an operation at this moment, we will interpret
+			// that as a remote runner, and this will return a false positive.
+
+			// Also note that this is designed to run our own CLI runner.
+			hasRemoteRunner = true
+			break
+		}
+	}
+	if !hasRemoteRunner {
+		log.Debug("No remote runner detected - operation cannot occur remotely")
+		return false, nil
+	}
+
+	// Check to see if we have a runner profile assigned to this project
+
+	if project.OndemandRunner != nil {
+		log.Debug("Project has an explicit ODR profile set - operation will work remotely")
+		return true, nil
+	}
+
+	// Check to see if we have a global default ODR profile
+
+	// TODO: it would be more efficient if we had an arg to filter to just get default profiles.
+	configsResp, err := client.ListOnDemandRunnerConfigs(ctx, &empty.Empty{})
+	if err != nil {
+		return false, err
+	}
+
+	defaultRunnerProfileExists := false
+	for _, odrConfig := range configsResp.Configs {
+		if odrConfig.Default {
+			defaultRunnerProfileExists = true
+			break
+		}
+	}
+
+	if defaultRunnerProfileExists {
+		log.Debug("Default runner profile exists - operation will work remotely.")
+		return true, nil
+	}
+
+	log.Debug("No runner profile is set for this project and no global default exists - operation should happen locally")
+	// The operation here _could_ still happen remotely - executed on the remote runner itself without ODR.
+	// If it's a container build op it will probably fail (because no kaniko), and if it's a deploy/release op it
+	// very well might fail do to incorrect/insufficient permissions. Because it probably won't work, we won't try,
+	// but the user could force it to happen locally by setting -local=false.
+	return false, nil
```

Comment:
Yeah this makes me wonder the difference between two semantics:

- `remoteIsPossible` (this function)
- `remoteIsPreferable` (hypothetical)

As @evanphx pointed out, this forces ODR completely, which isn't unreasonable, but will the CLI allow FORCING remote still? If so, no problem, but it still may be useful to have the `remoteIsPossible` validation which could return `true` even w/o ODR.

Reply:
Another thought is, rather than look at the code and hope it works, add a test job to the system that can be launched and watched. So that way we validate the pipeline with an empirical test rather than a theoretical one.

Reply:
We have the code for this already, we have noop jobs. So that's a good idea, just very slow, but maybe right during preflight for real ops.

Reply:
Yeah, it depends entirely when this code would be run. You wouldn't want to be doing it to decide if every operation could take place. I'd rather have a flag somewhere that says "this project has run remote jobs before" that is set in the past and this tool just checks that flag.

Reply:
The CLI will still be able to force a job happen remotely - so if they know something these heuristics don't (i.e. they want it to run on the remote runner with no ODR), they can still queue up the remote job. 

I like the idea of a test-baloon job to see of remote ops are possible, but for full coverage we'd need to run it before every op (too slow), or reevaluate it every time the project gets upserted, every time a runner profile is upserted, and every time a runner registers (without additional heuristics).

 I think this is probably good enough for now, but I can see doing that down the road.